sway-lib-std/src/reentrancy.sw

```diff
@@ -0,0 +1,42 @@
+//! A reentrancy check for use in Sway contracts.
+//! Note that this only works in internal contexts.
+//! to prevent reentrancy: `assert(!is_reentrant);
+
+library reentrancy;
+
+use ::context::call_frames::*;
+use ::constants::ZERO;
+use ::assert::assert;
+use ::panic::panic;
+use ::chain::auth::caller_is_external;
+use ::chain::log_u64;
+use ::option::*;
+use ::contract_id::ContractId;
+use ::context::registers::frame_ptr;
+
+pub fn reentrancy_guard() {
+    assert(is_reentrant() == false);
+}
+
+/// Returns `true` if the reentrancy pattern is detected, and `false` otherwise.
+pub fn is_reentrant() -> bool {
+    let mut reentrancy = false;
+    let mut call_frame_pointer = frame_ptr();
+    // Get our current contract ID
+    let this_id = contract_id();
+
+    // Reentrancy cannot occur in an external context.
+    // If not detected by the time we get to the bottom of the call_frame stack,
+    // then no reentrancy has occured.
+    while call_frame_pointer != 0 {
+        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
+        // get the ContractId value from the previous call frame
+        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
+        if previous_contract_id == this_id {
+            reentrancy = true;
```

Comment:
You could just `return true` here and not need a `reentrancy` bool at all.

Reply:
I will revisit this. There was a reason I chose this way, I think it was this error:

```
let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
33 |           if previous_contract_id == this_id {
   |  _________^
34 | |             true
35 | |             // reentrancy = true;
36 | |             // call_frame_pointer = 0;
37 | |         }
38 | |     }
   | |____^ Mismatched types.
expected: ()
found:    bool.
help: A while loop's loop body cannot implicitly return a value.Try assigning it to a mutable variable declared outside of the loop instead.
```

But I'll try with an explicit `return`.

Reply:
Yeah, it'd need to be explicit.

Reply:
So I tried to clean this up with an explicit return with no luck. Everything I tried either:
- caused tests to fail with `OutOfGas`, or
- caused tests to fail with `fatal runtime error: stack overflow`

Reply:
For reference, this overflows the stack:
```
pub fn is_reentrant() -> bool {
    let mut reentrancy = false;
    let mut call_frame_pointer = frame_ptr();
    let this_id = contract_id();
    while call_frame_pointer != 0 {
        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
        if previous_contract_id == this_id {
            call_frame_pointer = 0;
            return true;
        };
    }
    false
}
```
And this causes `OutOfGas` errors:
```
pub fn is_reentrant() -> bool {
    let mut reentrancy = false;
    let mut call_frame_pointer = frame_ptr();
    let this_id = contract_id();
    while call_frame_pointer != 0 {
        call_frame_pointer = get_previous_frame_pointer(call_frame_pointer);
        let previous_contract_id = get_contract_id_from_call_frame(call_frame_pointer);
        if previous_contract_id == this_id {
            return true;
        };
    }
    false
}
```

Reply:
#1128